src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    stateHistory.push(currentState);
```

Comment:
[DONT REPEAT YOURSELF] - The action of pushing the currentState to the stateHistory is repeated in every case of the switch or if/else. This action should be performed just once after the switch or if/else block. This will ensure that the code is not repetitive and follows the DRY principle.